stats/exporter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,242 @@
+	ProjectID string
+
+	// OnError is the hooked to be called when there is
+	// an error occured when uploading the stats data.
+	// If no custom hook is set, errors are logged.
+	// Optional.
+	OnError func(err error)
+
+	// ClientOptions are additional options to be passed
+	// to the underlying Stackdriver Monitoring API client.
+	// Optional.
+	ClientOptions []option.ClientOption
+
+	// ExportDelayThreshold determines the max amount of time
+	// the exporter can wait before uploading view data to
+	// the backend.
+	// Optional.
+	ExportDelayThreshold time.Duration
+
+	// ExportCountThreshold determines how many view data events
+	// can be buffered before batch uploading them to the backend.
+	// Optional.
+	ExportCountThreshold int
+
+	bundler *bundler.Bundler
+
+	measuresMu sync.Mutex
+	measures   map[stats.Measure]struct{} // measures already created remotely
+
+	once sync.Once
+	c    *monitoring.MetricClient
+}
+
+// Export exports to the Stackdriver Monitoroing if view data
+// has one or more rows.
+func (e *Exporter) Export(vd *stats.ViewData) {
+	e.once.Do(e.newClient)
+	if len(vd.Rows) == 0 {
+		return
+	}
+	e.bundler.Add(vd, 1)
+}
+
+func (e *Exporter) onError(err error) {
+	if e.OnError != nil {
+		e.OnError(err)
+		return
+	}
+	log.Printf("Failed to export to Stackdriver Monitoring: %v", err)
+}
+
+func (e *Exporter) newClient() {
+	client, err := monitoring.NewMetricClient(context.Background(), e.ClientOptions...)
+	if err != nil {
+		e.OnError(err)
+		return
+	}
+	e.c = client
+	e.measures = make(map[stats.Measure]struct{})
+	e.bundler = bundler.NewBundler((*stats.ViewData)(nil), func(bundle interface{}) {
+		vds := bundle.([]*stats.ViewData)
+		if err := e.upload(vds); err != nil {
+			e.onError(err)
+		}
+	})
+	e.bundler.DelayThreshold = e.ExportDelayThreshold
+	e.bundler.BundleCountThreshold = e.ExportCountThreshold
+}
+
+func (e *Exporter) upload(vds []*stats.ViewData) error {
+	ctx := context.Background()
+
+	// TODO(jbd): Create measures concurrently?
+	for _, vd := range vds {
+		if err := e.createMeasure(ctx, vd); err != nil {
+			return err
+		}
+	}
+	if err := e.c.CreateTimeSeries(ctx, e.makeReq(vds)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *Exporter) makeReq(vds []*stats.ViewData) *monitoringpb.CreateTimeSeriesRequest {
+	timeSeries := make([]*monitoringpb.TimeSeries, len(vds))
+	for i, vd := range vds {
+		timeSeries[i] = &monitoringpb.TimeSeries{
+			Metric: &metricpb.Metric{
+				Type: path.Join("custom.googleapis.com", vd.V.Measure().Name()),
+				// TODO(jbd): Add labels.
+			},
+			Resource: &monitoredrespb.MonitoredResource{
+				Type:   "global",
+				Labels: map[string]string{"project_id": e.ProjectID},
+			},
+			Points: newPoints(vd),
+		}
+	}
+	return &monitoringpb.CreateTimeSeriesRequest{
+		Name:       monitoring.MetricProjectPath(e.ProjectID),
+		TimeSeries: timeSeries,
+	}
+}
+
+func (e *Exporter) createMeasure(ctx context.Context, vd *stats.ViewData) error {
```

Comment:
Is `createMeasure` called concurrently?  I'm trying to reason through possible contention, since this makes an RPC while holding the mutex.